sdks/go/pkg/beam/core/runtime/exec/util.go

```diff
@@ -33,13 +34,29 @@ func (g *GenID) New() UnitID {
 	return UnitID(g.last)
 }
 
+type doFnError struct {
+	doFn string
+	err  error
+	uid  UnitID
+	pid  string
+}
+
+func (e *doFnError) Error() string {
+	return fmt.Sprintf("DoFn[<%d>;<%s>]<%s> returned error:<%s>", e.uid, e.pid, e.doFn, e.err)
+}
+
 // callNoPanic calls the given function and catches any panic.
 func callNoPanic(ctx context.Context, fn func(context.Context) error) (err error) {
 	defer func() {
 		if r := recover(); r != nil {
-			// Top level error is the panic itself, but also include the stack trace as the original error.
-			// Higher levels can then add appropriate context without getting pushed down by the stack trace.
-			err = errors.SetTopLevelMsgf(errors.Errorf("panic: %v %s", r, debug.Stack()), "panic: %v", r)
+			// Check if the panic value is from a failed DoFn, and return it without a pancic trace.
+			if e, ok := r.(doFnError); ok {
```

Comment:
I don't think this line runs as expected: in the fail() call in pardo.go it's passed a *doFnError to the guarded error, but then it's type asserting the recovery value here as the non-pointer type.  And then in the _wrappedPanic case, a non-pointer doFnError is used instead of the pointer, not replicating what's happening in fail.

Reply:
Yes, I missed this.